server/data/metrics.ts

```diff
@@ -0,0 +1,84 @@
+import { cycles } from '../config/mongoCollections';
+import { ObjectId } from 'mongodb';
+import { checkObjectId } from '../helpers/error';
+import { CycleObject, MetricsObject } from '../typings';
+
+/**
+ * @description Gets metric data for cycle with given id
+ * @param {string} id Cycle id
+ * @returns {Promise<MetricsObject>} Metrics object
+ */
+export const getMetricsByID = async (id: string): Promise<MetricsObject> => {
+    if(!checkObjectId(id)){
+        throw new Error('Invalid id');
+    }
+
+    const cycleCollection = await cycles();
+    const cycle: CycleObject = await cycleCollection.findOne({
+        _id: new ObjectId(id)});
+    if(cycle === null)
+        throw new Error("There is no cycle with that id.");
+
+    // Set up variables for metrics object
+    let num_saved: number = 0;
+    let num_applications: number = 0;
+    let num_interviewed: number = 0;
+    let num_offers: number = 0;
+    let num_rejections: number = 0;
+    let num_rounds: number = 0;
+    let avg_salary: number = 0;
+    let num_connections: number = 0;
+    let application_timeline: Array<Date> = [];
+
+    let salaries: Array<number> = [];
+
+    for(let app of cycle.applications) { // For each application
+        num_saved++; // another saved application
+        let interview_flag = false;
+
+        for(let event of app.events) { // For each event
+            if(event.title === "Apply" && event.status === true) {
+                num_applications++; // another application
+                application_timeline.push(event.date); // add to timeline
+            } else if(event.title !== "Saved" && event.status === true) {
+                if(!interview_flag) {
+                    num_interviewed++; // another interviewed application
+                    interview_flag = true;
+                }
+                num_rounds++; // another round
+            }
+        }
```

Comment:
num_application represents number applied. I'll remove "saved" since it doesn't make sense as a title since every application is implied to be saved by the nature of its existence. All the titles can be whatever the user puts (which is why we might want to add a type of event to the event object), but for now, we'll auto make the first event of a saved application to be "Apply", and then, treat every other event as an interview round.